### ISSUE TITLE
Inconvenient MockRef<'a> API

### ISSUE THREAD

**Issue by gdesmott** [alexliesenfeld/httpmock#26]
Hi,

I'm trying to write a non-trivial mock server. To do so I'd need to store both the `MockServer` object and a bunch of `MockRef<'a>` in the same struct (the one representing my mock server). Unfortunately it's pretty hard to do as `MockRef` is itself keeping a ref on `Server`, introducing the well-known self-referential problem.

It is possible to workaround it with tricks such as [rental](https://crates.io/crates/rental) or [owning_ref](https://crates.io/crates/owning_ref) but it greatly increase the code complexity.
An easier approach would be to remove the server borrow in  `MockRef` by wrapping it around a `Rc` or `Arc`.
Would you be opened to such change?

**Comment by alexliesenfeld** [comment]
Hi,

the idea behind this was to model the fact that a mock object can never outlive the mock server it was created on. I feel that removing this constraint would have some negative impact on developer experience. 

However, I think it could be enough to expose the `id` attribute from the `MockRef` and add a constructor function `new` to `MockRef` that could be used this way:

```
// ******************************** Your library code - START *****************************
struct CustomMockRef {
    id: usize
}

struct MyWrapper {
    server: MockServer,
    mocks: RefCell<Vec<CustomMockRef>>
}

 let w = MyWrapper {
    server: MockServer::start(),
    mocks: RefCell::new(vec![])
 };
// ******************************** Your library code - END *******************************

let mock = w.server.mock(|when, then| {
    when.path("/test");
    then.status(200);
});

// ******************************** Your library code - START *****************************
w.mocks.borrow_mut().push(CustomMockRef{id: mock.id});
// ******************************** Your library code - END *******************************

let mock : MockRef = MockRef::new(mock.id, &w.server);
mock.assert();
``` 

This way you could recreate the `MockRef` structure yourself using the `id` and a `MockServer` yourself. 

Would that be sufficient to you?

**Comment by gdesmott** [comment]
Thanks for your quick reply Alex.

Yes, such change would solve my problem as well.
From an API pov I'm wondering if having `MockServer::fetch_mock(&self, id: usize) -> Result<MockRef, _>`
wouldn't be cleaner?

**Comment by alexliesenfeld** [comment]
The `MockServer::fetch_mock` would not actually fetch anything from the `MockServer` but only create a new `MockRef` instance. Since we are dealing with recreating a mock reference, I think it should go into the `MockRef` structure. 

I have created an initial implementation for this in #28. The code that is interesting for you is within file `src/api/mock.rs` (see [here](https://github.com/alexliesenfeld/httpmock/pull/28/files#diff-435c6064639170f0d24616bc1f5df9abcdcebd1364405705064d483a39a95147R390-R448)). 

Comments are very welcome!

**Comment by gdesmott** [comment]
I just tested the alpha release and it works great for my use cases.
Thanks a lot @alexliesenfeld  !

**Comment by alexliesenfeld** [comment]
Very welcome :-)

**Comment by alexliesenfeld** [comment]
Has been released in [v0.5.4](https://github.com/alexliesenfeld/httpmock/releases/tag/v0.5.4)
